### PR TITLE
added rfc for node identity

### PR DESCRIPTION
**Pull request by woutslakhorst**
part of #137.

I'm not sure where to describe the encrypted `to` header for private DAG payloads. This might be part of the V2 protocol? Or does it need it's own RFC?

### CODE REVIEW

File: rfc/rfc015-node-identity.md [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+|  | Nov 2021 |
+
+## Node identity
+
+### Abstract
+
+This RFC describes how to identify a node in the network. Node identification is required for exchanging private data.
+Measures to keep data private can range from encrypting data to sending data over a specific connection.
+The TLS connection and certificate exchange form the basis of the node identification.
+The *Subject Alternative Name* of the certificate is matched against the *serviceEndpoint* of a DID Document service.
```
Comment (dmeijboom) [inline rfc/rfc015-node-identity.md:16]: > The *Subject Alternative Name* of the (...)

It might be good to specify which one or all of them as you can have multiple SANs.

Comment (dmeijboom) [reply rfc/rfc015-node-identity.md:16]: We might want to add limitations for the SAN as they can contain an e-mail address, DNS name, IP address or URI.

File: rfc/rfc015-node-identity.md [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+### Copyright Notice
+
+![](../.gitbook/assets/license.png)
+
+This document is released under the [Attribution-ShareAlike 4.0 International \(CC BY-SA 4.0\) license](https://creativecommons.org/licenses/by-sa/4.0/).
+
+## 1. Introduction
+
+When all participants of a network share the same public data, there's no need to know which node services a particular organization.
+This changes when data becomes private. Private data needs to be routed to the correct node.
```
Comment (dmeijboom) [inline rfc/rfc015-node-identity.md:38]: > routed to the correct node (...)


Send to the correct node right? As "routed" sounds like we have some sort of routing protocol or something like that

File: rfc/rfc015-node-identity.md [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+In order for that to work, there's a need to identify each node and link that identity to services organizations.
+[RFC014](rfc014-authorization-credential.md) describes that Nuts authorization credentials need to remain private.
```
Comment (dmeijboom) [inline rfc/rfc015-node-identity.md:40]: The last sentence feels a little bit 'off' here. Maybe move it to a section that explains more about private data?

File: rfc/rfc015-node-identity.md [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+Two things need to be resolved:
+
+1. Node identity for peers that are connected. Network connections are the only interaction a node has to another node, so the connecting peer needs to be identified.
+2. DIDs need to be linked to DIDs that have a node identity.
```
Comment (dmeijboom) [inline rfc/rfc015-node-identity.md:45]: Maybe rephrase this? I'm not sure what this means.

**Review by dmeijboom — CHANGES_REQUESTED** [review]

File: rfc/rfc015-node-identity.md (hunk 1 above)
Comment (woutslakhorst) [reply rfc/rfc015-node-identity.md:16]: I will add the particulars

File: rfc/rfc015-node-identity.md (hunk 2 above)
Comment (woutslakhorst) [reply rfc/rfc015-node-identity.md:38]: Yes, that may be added to a future RFC.

File: rfc/rfc015-node-identity.md (hunk 3 above)
Comment (woutslakhorst) [reply rfc/rfc015-node-identity.md:40]: I removed that sentence

**Review by dmeijboom — APPROVED** [review]

File: rfc/rfc015-node-identity.md [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+TLS certificates are linked to a domain name via a SAN extension.
+
+```
+X509v3 extensions:
+  X509v3 Subject Alternative Name:
+    DNS:example.com
+```
+
+This domain is checked by a client that connects to a server with that domain. The certificate the server uses must match the domain the client dialed.
```
Comment (reinkrul) [inline rfc/rfc015-node-identity.md:69]: this is TLS 101 (hostname check) and should be considered standard practice? Or is it just to inform the reader on how it works in general?

**Review by reinkrul — CHANGES_REQUESTED** [review]

File: rfc/rfc015-node-identity.md (hunk 5 above)
Comment (woutslakhorst) [reply rfc/rfc015-node-identity.md:69]: Just to remind that there's a difference between being a client and a server.

File: rfc/rfc015-node-identity.md [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+  "serviceEndpoint": "grpc://example.com:5555"
+}
+```
+
+The `type` MUST be `NutsComm`. 
+The `serviceEndpoint` MUST start with `grpc://` and end with the correct port number, `:5555` in this example.
+
+### 3.3 Connecting identity
+
+To determine the connecting identity, a `DID` metadata header SHOULD be sent by each node upon establishing a connection.
```
[the remark this answers is not in the document]
Comment (woutslakhorst) [reply rfc/rfc015-node-identity.md:92]: There's no specific concept of a `nodeDID`, there's only the DID that specifies the `NutsComm` service.

File: rfc/rfc015-node-identity.md [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+```
+{
+  "id": "did:nuts:123#IyvdTGF1Dsgwngfdg3SH6TpDv0Ta1aOEkw",
+  "type": "NutsComm",
+  "serviceEndpoint": "grpc://example.com:5555"
+}
+```
+
+The `type` MUST be `NutsComm`. 
+The `serviceEndpoint` MUST start with `grpc://` and end with the correct port number, `:5555` in this example.
```
[the remark this answers is not in the document]
Comment (woutslakhorst) [reply rfc/rfc015-node-identity.md:88]: added it here

File: rfc/rfc015-node-identity.md [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+
+### Abstract
+
+This RFC describes how to identify a node in the network. Node identification is required for exchanging private data.
+Measures to keep data private can range from encrypting data to sending data over a specific connection.
+The TLS connection and certificate exchange form the basis of the node identification.
+The *Subject Alternative Name* of the certificate is matched against the *serviceEndpoint* of a DID Document service.
+The requirement for the additional service in the DID Document also makes it possible to find other nodes and thus enable service discovery of Nuts nodes.
+
+Private data, like authorization credentials, have DIDs as issuer and subject. To enable private addressing, a DID will be encrypted with its public encryption key.
```
[the remark this answers is not in the document]
Comment (woutslakhorst) [reply rfc/rfc015-node-identity.md:19]: There's too much info on private addressing, which is a separate RFC. Will rewrite

[the remark this answers is not in the document]
Comment (woutslakhorst) [reply rfc/rfc015-node-identity.md:19]: I'll add a link to the correct RFC for private addressing when available.

**Review by reinkrul — APPROVED** [review]
